### ISSUE TITLE
No Daisy-Chaining .concat()

### ISSUE THREAD

**Issue by JonathanWilbur** [eslint/eslint#13206]
**Please describe what the rule should do:**

Prohibit / warn against daisy-chaining of calls to `Array.concat()`. It is more efficient to use a single call to `.concat()` supplying all of the arrays as parameters to it, because the ECMAScript engine can calculate the ultimate length of the array one time and allocate the appropriate amount of memory, rather than repeatedly recalculating and reallocating.

**What category of rule is this? (place an "X" next to just one item)**

[x] Warns about a potential error (problem)
[x] Suggests an alternate way of doing something (suggestion)
[x] Enforces code style (layout)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

Bad:

```js
const a = [ 'a', 'b', 'c' ];
a.concat([ 'd', 'e', 'f' ]).concat([ 'g', 'h', 'i' ]);
```

Good:

```js
const a = [ 'a', 'b', 'c' ];
a.concat([ 'd', 'e', 'f' ], [ 'g', 'h', 'i' ]);
```

**Why should this rule be included in ESLint (instead of a plugin)?**

It is applicable to almost all versions of ESLint and there is almost no reason you would not want to use it.

**Are you willing to submit a pull request to implement this rule?**

No.

**Comment by anikethsaha** [comment]
it seems like a good feature. it needs a new rule as I think there are no rules currently working in this scope. 
Maybe this can fit better for [`eslint-plugin-array-func`](https://github.com/freaktechnik/eslint-plugin-array-func) ?

**Comment by kaicataldo** [comment]
I agree with @anikethsaha that this feels a like a better fit for a custom rule. Since ESLint doesn't have access to type information, we would not be able to implement this without false negatives (i.e. any object that has as `.concat()` method that isn't an array).
